compiler/resolution/callDestructors.cpp

```diff
@@ -1278,29 +889,49 @@ static void insertYieldTemps()
       continue;
 
     // This is the symbol passed back in the yield.
-    SymExpr* yieldExpr = toSymExpr(call->get(1));
+    SymExpr* yieldedSe = toSymExpr(call->get(1));
+    Symbol* yieldedSym = yieldedSe->symbol();
 
-    // The transformation is applied only if is has a normal record type
-    // (passed by value).
-    Type* type = yieldExpr->symbol()->type;
+    FnSymbol* inFn = toFnSymbol(call->parentSymbol);
+    IteratorInfo* ii = inFn->iteratorInfo;
 
-    if (isRecord(type) &&
-        !type->symbol->hasFlag(FLAG_ITERATOR_RECORD) &&
-        !type->symbol->hasFlag(FLAG_RUNTIME_TYPE_VALUE))
-    {
-      SET_LINENO(call);
+    // coforall functions in iterators don't seem to have ii
+    RetTag iteratorRetTag = RET_VALUE;
+    if (ii)
+      iteratorRetTag = ii->iteratorRetTag;
 
-      // Replace:
-      //   yield <yieldExpr>
-      // with:
-      //   (def _yield_expr_tmp_:type)
-      //   (move _yield_expr_tmp_ ("chpl__autoCopy" <yieldExpr>))
-      //   yield _yield_expr_tmp_
-      Symbol* tmp = newTemp("_yield_expr_tmp_", type);
-      Expr* stmt = call->getStmtExpr();
-      stmt->insertBefore(new DefExpr(tmp));
-      stmt->insertBefore(new CallExpr(PRIM_MOVE, tmp, new CallExpr(getAutoCopyForType(type), yieldExpr->remove())));
-      call->insertAtHead(new SymExpr(tmp)); // New first argument.
+
+    // If the yielded symbol is subject to auto-copy/destroy discipline
+    // and it's returned by value (not by reference)
+    // and the yielded value is not an expression temporary
+    //  (e.g. for yield someCall(), the result of someCall() doesn't need copy)
```

Comment:
Seems like the opening "If" of this sentence is lacking "then". Would you double-check?